src/metabase/lib/drill_thru/zoom_in_timeseries.cljc

```diff
@@ -48,13 +50,26 @@
    [metabase.shared.util.i18n :as i18n]
    [metabase.util.malli :as mu]))
 
-;;; TODO -- we shouldn't include hour and minute for `:type/Date` columns.
-(def ^:private valid-current-units
-  [:year :quarter :month :week :day :hour :minute])
+(mu/defn- valid-current-units-for-schema-type :- [:vector :keyword]
```

Comment:
We could use `lib.temporal-bucket/available-temporal-buckets` here. It returns all available options for per column type; but it seems that we need truncation units only. We could use `lib.schema.temporal-bucketing/datetime-truncation-units` for this. Maybe something like:
```
(into []
  (filter (fn [option]
            (->> (lib.temporal-bucket/raw-temporal-bucket option)
                 (contains? lib.schema.temporal-bucketing/datetime-extraction-units))))
  (lib.temporal-bucket/available-temporal-buckets query stage-number column))
```

As a bonus, it:
1. Actually handles `effective-type` correctly
2. Handles custom expressions for which we don't allow temporal buckets in breakouts

Reply:
Hi @ranquild, thanks for the review!

I'm glad you mention `lib.temporal-bucket` and `:effective-type` as these were two things I was going to ask about specifically.

In particular, I don't use `available-temporal-buckets` because it is checking `:effictive-type` which turns out to be `:type/Integer` when bucketing by `:year`.

``` clojure
(-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
    (lib/breakout (-> (meta/field-metadata :orders :created-at)
                      (lib/with-temporal-bucket :year)))
    :stages
    first
    :breakout)
==>
[[:field
  {:lib/uuid "d2588602-8018-42d2-92f6-83fb8e7e392b",
   :base-type :type/DateTimeWithLocalTZ,
   :effective-type :type/Integer,
   :temporal-unit :year,
   :metabase.lib.field/original-effective-type :type/DateTimeWithLocalTZ}
  57606]]
```

And, consequently:

``` clojure
(let [query (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
                (lib/aggregate (lib/count))
                (lib/breakout (-> (meta/field-metadata :orders :created-at)
                                  (lib/with-temporal-bucket :year))))
      option {:lib/type :metadata/column,
              :base-type :type/DateTimeWithLocalTZ,
              :effective-type :type/Integer}]
  (lib.temporal-bucket/available-temporal-buckets query -1 option))

==>

[]
```

See, e.g. the comment at the top of

``` clojure
(defmethod lib.temporal-bucket/with-temporal-bucket-method :field
  [[_tag options id-or-name] unit]
  ;; if `unit` is an extraction unit like `:month-of-year`, then the `:effective-type` of the ref changes to
  ;; `:type/Integer` (month of year returns an int). We need to record the ORIGINAL effective type somewhere in case
  ;; we need to refer back to it, e.g. to see what temporal buckets are available if we want to change the unit, or if
  ;; we want to remove it later. We will record this with the key `::original-effective-type`. Note that changing the
  ;; unit multiple times should keep the original first value of `::original-effective-type`.
  (if unit
    (let [extraction-unit?        (contains? lib.schema.temporal-bucketing/datetime-extraction-units unit)
          original-effective-type ((some-fn ::original-effective-type :effective-type :base-type) options)
		  ...
```

The reason being that `lib.schema.temporal-bucketing/datetime-extraction-units` contains `:year`.

That is why I check for `original-effective-type` and `base-type` rather than `effective-type`, but I wanted to ask about that as I was unsure whether that was the right approach...

Also thanks for the tip about filtering `available-temporal-buckets`. I had seen `lib.schema.temporal-bucketing/datetime-truncation-units` but noticed that it was (a) unordered (there is an `ordered-date-truncation-units` but no corresponding `ordered-datetime-truncation-units`, though I guess it would be easy to add) and (b) contains "extra" units like `:second` and `:millisecond` that the old code here didn't include. But I see now that filtering `available-temporal-buckets` by `datetime-truncation-units` seems to be exactly what we'd want, except for the case of `:year` buckets mentioned above.

Reply:
it's usually incorrect to check for either `base-type` or `effective-type` manually. There are many cases to be aware of, like coercion strategies, buckets changing the type etc.

Reply:
are you sure that the behavior you want to achieve with `:year` is correct?

Reply:
@bshepherdson please take a look

Reply:
I think the right way to do this is using `available-temporal-buckets-method` on the breakout ref. That refers to the type of the original field, ignoring any bucketing, which should be `Date`, `DateTime`, or `Time`, and return the right set.

So I'd approach it like this:
- `lib.temporal-bucket/available-temporal-buckets` to get the list for that column
- create `lib.schema.temporal-bucketing/ordered-datetime-truncation-units` since it doesn't exist yet
- filter the `ordered-datetime-truncation-units` to only those which are `available`
- the next one in that list is the one!

Reply:
@bshepherdson how do I get the breakout ref? I'm guessing it's something other than the thing I'm selecting in my example above, but I didn't see anything in the query or context that is passed in that looks like a reference to a breakout with the original field's `effective-type`.

In other words, not this, I assume:

```clojure
(-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
    (lib/breakout (-> (meta/field-metadata :orders :created-at)
                      (lib/with-temporal-bucket :year)))
    :stages
    first
    :breakout)
==>
[[:field
  {:lib/uuid "d2588602-8018-42d2-92f6-83fb8e7e392b",
   :base-type :type/DateTimeWithLocalTZ,
   :effective-type :type/Integer,
   :temporal-unit :year,
   :metabase.lib.field/original-effective-type :type/DateTimeWithLocalTZ}
  57606]]
```

Or can I use that uuid to look it up?

Reply:
That's the correct ref. The right way to access it is with `(first (lib.breakout/breakouts query -1))`, though. Disregard its `:effective-type`; `available-temporal-buckets` should do the right thing anyway.

(It looks up the original input field rather than relying on the breakout directly, see the `(defmethod available-temporal-buckets-method :field ...)` in `lib.field`.)

Reply:
For context, this nonsense is the kind of thing I mean when I said in the retro this morning that refs suck and are the source of much of our pain, bugs and perf waste.

In the prototype I mentioned, breakouts are themselves columns and contain a reference (the ID of another entity in the "database"; which is to say a real, tangible reference rather than a symbolic label!) so you can navigate to the input column with a `get`: `(:mbql.breakout/column a-breakout)` and done.

Reply:
Yep, that works! I will rework to use this method instead. Thanks!

```clojure
(let [query (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
                (lib/aggregate (lib/count))
                (lib/breakout (-> (meta/field-metadata :orders :created-at)
                                  (lib/with-temporal-bucket :year))))
      breakout (first (lib.breakout/breakouts query -1))]
  (into []
        (comp 
         (map lib.temporal-bucket/raw-temporal-bucket)
         (filter #(contains? lib.schema.temporal-bucketing/datetime-truncation-units %)))
        (lib.temporal-bucket/available-temporal-buckets query -1 breakout)))

==>

[:minute :hour :day :week :month :quarter :year]
```